src/state_migration/nv17/miner.rs

```diff
@@ -0,0 +1,925 @@
+                                let out_sectors_snapshot_cid = out_sectors_snapshot.flush()?;
+                                cache.insert(
+                                    out_sectors_snapshot_cid_cache_key,
+                                    out_sectors_snapshot_cid,
+                                );
+                                out_sectors_snapshot_cid
+                            }
+                        };
+
+                    let out_deadline = fil_actor_miner_state::v9::Deadline {
+                        partitions: in_deadline.partitions,
+                        expirations_epochs: in_deadline.expirations_epochs,
+                        partitions_posted: in_deadline.partitions_posted,
+                        early_terminations: in_deadline.early_terminations,
+                        live_sectors: in_deadline.live_sectors,
+                        total_sectors: in_deadline.total_sectors,
+                        faulty_power: TypeMigrator::migrate_type(in_deadline.faulty_power, store)?,
+                        optimistic_post_submissions: in_deadline.optimistic_post_submissions,
+                        sectors_snapshot: out_sectors_snapshot_cid,
+                        partitions_snapshot: in_deadline.partitions_snapshot,
+                        optimistic_post_submissions_snapshot: in_deadline
+                            .optimistic_post_submissions_snapshot,
+                    };
+
+                    let out_deadline_cid = store.put_cbor_default(&out_deadline)?;
+
+                    if i < out_deadlines.due.len() {
+                        out_deadlines.due[i] = out_deadline_cid;
+                    } else {
+                        out_deadlines.due.push(out_deadline_cid);
+                    }
+                }
+            }
+
+            store.put_cbor_default(&out_deadlines)
+        }
+    }
+}
+
+fn migrate_from_scratch<'bs, BS: Blockstore>(
+    store: &'bs BS,
+    in_array: &fil_actors_shared::v8::Array<fil_actor_miner_state::v8::SectorOnChainInfo, BS>,
+) -> anyhow::Result<
+    fil_actors_shared::v9::Array<'bs, fil_actor_miner_state::v9::SectorOnChainInfo, BS>,
+> {
+    use fil_actor_miner_state::v9::SECTORS_AMT_BITWIDTH;
+
+    let mut out_array = fil_actors_shared::v9::Array::<
+        fil_actor_miner_state::v9::SectorOnChainInfo,
+        _,
+    >::new_with_bit_width(store, SECTORS_AMT_BITWIDTH);
+
+    in_array.for_each(|key, info_v8| {
+        out_array.set(key, TypeMigrator::migrate_type(info_v8.clone(), store)?)?;
+        Ok(())
+    })?;
+
+    Ok(out_array)
+}
+
+fn miner_prev_sectors_in_key(addr: &Address) -> String {
+    format!("prevSectorsIn-{addr}")
+}
+
+fn miner_prev_sectors_out_key(addr: &Address) -> String {
+    format!("prevSectorsOut-{addr}")
+}
+
+fn sectors_amt_key(cid: &Cid) -> anyhow::Result<String> {
+    Ok(format!(
+        "sectorsAmt-{}",
+        cid.to_string_of_base(Base::Base32Lower)?,
+    ))
+}
```

Comment:
Let's format the strings in a `rust_way`.

Reply:
Done.